src/realm.h

```diff
@@ -4069,18 +4069,37 @@ RLM_API void realm_register_user_code_callback_error(realm_userdata_t usercode_e
 typedef struct realm_thread_observer_token realm_thread_observer_token_t;
 
 /**
- * Register a callback handler for bindings interested in registering callbacks before/after the ObjectStore thread
- * runs.
+ * Register a global callback handler for bindings interested in registering callbacks before/after
+ * the ObjectStore thread runs. There can only be one global callback handler registered at a time.
+ * Call realm_release() on the returned token pointer before assigning a new set of callback handlers.
+ * This only works for the default socket provider implementation.
  * @param on_thread_create callback invoked when the object store thread is created
  * @param on_thread_destroy callback invoked when the object store thread is destroyed
  * @param on_error callback invoked to signal to the listener that some error has occured.
+ * @param user_data pointer to user defined data that is provided to each of the callback functions
+ * @param free_userdata callback invoked when the user_data is to be freed
  * @return a token that has to be released in order to stop receiving notifications
  */
 RLM_API realm_thread_observer_token_t*
-realm_set_binding_callback_thread_observer(realm_on_object_store_thread_callback_t on_thread_create,
-                                           realm_on_object_store_thread_callback_t on_thread_destroy,
-                                           realm_on_object_store_error_callback_t on_error, realm_userdata_t,
-                                           realm_free_userdata_func_t free_userdata);
+realm_set_global_binding_thread_observer(realm_on_object_store_thread_callback_t on_thread_create,
+                                         realm_on_object_store_thread_callback_t on_thread_destroy,
+                                         realm_on_object_store_error_callback_t on_error, realm_userdata_t user_data,
+                                         realm_free_userdata_func_t free_userdata);
+
+/**
+ * Register an app local callback handler for bindings interested in registering callbacks before/after
+ * the ObjectStore thread runs for this app. This only works for the default socket provider implementation.
+ * @param config SyncClientConfig ptr created by realm_sync_client_config_new()
+ * @param on_thread_create callback invoked when the object store thread is created
+ * @param on_thread_destroy callback invoked when the object store thread is destroyed
+ * @param on_error callback invoked to signal to the listener that some error has occured.
+ * @param user_data pointer to user defined data that is provided to each of the callback functions
+ * @param free_userdata callback invoked when the user_data is to be freed
+ */
+RLM_API void realm_sync_client_config_set_default_binding_thread_observer(
```

Comment:
are both functions needed? Isn't this one enough?

Reply:
One sets the global thread observer (the way it works now) and the other sets the thread observer in `SyncClientConfig` to specify one on a "per app" basis (the new way).

Reply:
Do we want to keep using it the way it works now (with the global thread observer)?

Reply:
I think we should keep both for now, and, if the Java/Kotlin SDK team decides to move one way or the other, we can remove the other method. I'm not sure if/when they will be ready to migrate to the app-specific usage.

Reply:
I thought we decided to just have the per-app version...

Reply:
This is the (only) function for setting the thread observer in the SyncClientConfig. If you prefer, I can move this function declaration so it is with the other `realm_sync_client_config_...` functions